nostr/relay.py

```diff
@@ -54,42 +51,55 @@ def __post_init__(self):
             on_error=self._on_error,
             on_close=self._on_close
         )
+        self._connection_thread: Thread = None
 
-    def connect(self):
-        self.ws.run_forever(
-            sslopt=self.ssl_options,
-            http_proxy_host=self.proxy_config.host if self.proxy_config is not None else None, 
-            http_proxy_port=self.proxy_config.port if self.proxy_config is not None else None,
-            proxy_type=self.proxy_config.type if self.proxy_config is not None else None,
-        )
+    def connect(self, is_reconnect=False):
+        if not self.is_connected():
+            with self.lock:
+                self._connection_thread = Thread(
+                    target=self.ws.run_forever,
+                    kwargs={
+                        "sslopt": self.ssl_options,
+                        "http_proxy_host": self.proxy_config.host if self.proxy_config is not None else None,
+                        "http_proxy_port": self.proxy_config.port if self.proxy_config is not None else None,
+                        "proxy_type": self.proxy_config.type if self.proxy_config is not None else None
+                    },
+                    name=f"{self.url}-connection"
+                )
+                self._connection_thread.start()
+
+            if not is_reconnect:
+                Thread(
+                    target=self.outgoing_messages_worker,
+                    name=f"{self.url}-outgoing-messages-worker",
+                    daemon=True
+                ).start()
 
+            time.sleep(1)
+        
     def close(self):
-        self.ws.close()
+        if self.is_connected():
+            self.ws.close()
 
-    def check_reconnect(self):
-        try:
-            self.close()
-        except:
-            pass
-        self.connected = False
-        if self.reconnect:
-            time.sleep(1)
-            self.connect()
+    def is_connected(self) -> bool:
+        with self.lock:
+            if self._connection_thread is None or not self._connection_thread.is_alive():
+                return False
+            else:
+                return True
 
     def publish(self, message: str):
-        self.queue.put(message)
+        self.outgoing_messages.put(message)
         
-    def queue_worker(self):
+    def outgoing_messages_worker(self):
         while True:
-            if self.connected:
-                message = self.queue.get()
+            if self.is_connected():
+                message = self.outgoing_messages.get()
                 try:
                     self.ws.send(message)
                     self.num_sent_events += 1
                 except:
-                    self.queue.put(message)
-            else:
-                time.sleep(0.1)
+                    self.outgoing_messages.put(message)
```

Comment:
Eventually more robust error-handling is probably needed here. If the message is just undeliverable for some reason, it's pointless to return it to the Queue. Maybe for now at least dump the traceback so we're aware of the failure. Something like:
```
except:
    import traceback
    traceback.print_tb()
```
(I forget which traceback calls do what w/formatting, presenting, etc)

Also can't recall if prints from within threads always make it out to the console.

Reply:
I think I disagree on the return into the Queue. Relays constantly drop connections and allow reconnect. This ensures that a message in the queue will be delivered when the connection is back up.

Since we keep track of how often an error was encountered (with `self.error_counter`) we should do something with it. Right now it just stops reconnecting after it reaches `self.error_threshold` but an exponentially increasing reconnect sleep timer would be more elegant IMO.